runner/runner/main.py

```diff
@@ -181,67 +185,129 @@ def load_tests(config: Mapping[str, Any]) -> None:
-            BUILD_WITH_PSMV="0",
-            BUILD_WITH_PSVR="0",
-            BUILD_WITH_OPENHMD="0",
-            BUILD_WITH_VIVE="0",
-            ILLIXR_PATH=str(runtime_path),
-            **monado_config,
-        ),
-    )
-    cmake(
-        openxr_app_path,
-        openxr_app_path / "build",
-        dict(CMAKE_BUILD_TYPE=cmake_profile, **openxr_app_config),
-    )
-    build_runtime(config, "so")
-    plugin_paths = threading_map(
-        lambda plugin_config: build_one_plugin(config, plugin_config),
-        [plugin_config for plugin_group in config["plugin_groups"] for plugin_config in plugin_group["plugin_group"]],
-        desc="Building plugins",
+        monado_build_opts,
+        env_override=env_monado,
     )
 
+    if not "openxr_app" in config["action"]:
+        raise RuntimeError(f"Missing 'openxr_app' property for action '{action_name}")
+
+    openxr_app_obj    : Mapping[str, Any] = config["action"]["openxr_app"]
+    openxr_app_config : Mapping[str, str] = openxr_app_obj.get("config", {})
+
+    openxr_app_path     : Optional[Path] # Forward declare type
+    openxr_app_bin_path : Path           # Forward declare type
+
+    if "src_path" in openxr_app_obj["app"]:
+        ## Pathify 'src_path' for compilation
+        openxr_app_path     = pathify(openxr_app_obj["app"]["src_path"], root_dir, cache_path, True , True)
+        openxr_app_bin_path = openxr_app_path / openxr_app_obj["app"]["bin_subpath"]
+    else:
+        ## Get the full path to the 'app' binary
+        openxr_app_path     = None
+        openxr_app_bin_path = pathify(openxr_app_obj["app"], root_dir, cache_path, True, True)
+
+    ## Compile the OpenXR app if we received an 'app' with 'src_path'
+    if openxr_app_path:
+        cmake(
+            openxr_app_path,
+            openxr_app_path / "build",
+            dict(CMAKE_BUILD_TYPE=cmake_profile, **openxr_app_config),
+        )
+
+    if not openxr_app_bin_path.exists():
+        raise RuntimeError(f"{action_name} Failed to build openxr_app (mainline={is_mainline}, path={openxr_app_bin_path})")
+
+    if is_mainline:
+        monado_target_name : str  = "monado-service"
+        monado_target_dir  : Path = monado_path / "build" / "src" / "xrt" / "targets" / "service"
+        monado_target_path : Path = monado_target_dir / monado_target_name
+
+        if not monado_target_path.exists():
+            raise RuntimeError(f"[{action_name}] Failed to build monado (mainline={is_mainline}, path={monado_target_path})")
+
+        env_popen: Mapping[str, str] = dict(**os.environ, **env_monado)
+
+        ## Open the Monado service application in the background
+        monado_socket = Path("/run/user/1000/monado_comp_ipc")
+        if monado_socket.exists():
+            ## The Monado service will crash if a previous instance is left behind
+            monado_socket.unlink()
+        monado_service_proc = subprocess.Popen([str(monado_target_path)], env=env_popen, cwd=Path())
```

Comment:
Reviewers should test `configs/monado-mainline.yaml`, checking for a successful link between `monado-service` and the XR application. The process appears to leave behind a sticky instance of `monado-service` that prevents new runs, and does not respond to the cleanup on Line 303.